api/include/opentelemetry/trace/span.h

```diff
@@ -135,6 +136,18 @@ class Span
                        attributes.begin(), attributes.end()});
   }
 
+  /** Adds a Link to newly created Span
+   *  @param spanContext the context of the linked span
+   *  @param attributes Link attributes
+   */
+
+  virtual void AddLink(const trace::Link &link) noexcept = 0;
```

Comment:
Just thinking about two scenarios which makes it different from Events, and having `trace::Link` would be helpful.
 - Creating a link, and then associating it with multiple spans. It would be helpful to have link as an entity in that case.
 -  Link should be associated with the Span during it's creation, and not afterwards. As per the specification 
     "Links cannot be added after Span creation." With link type, it's helpful to pass sequence of links in `Span` constructor. Currently we do provide `Span::AddLink()` but it needs to be called immediately after the span creation ( probably this needs to be clearly documented).

Reply:
> Links cannot be added after Span creation.

I see. What do you think about completely dismissing `AddLink` methods, and just adding links via [StartSpanOptions](https://github.com/open-telemetry/opentelemetry-cpp/blob/master/api/include/opentelemetry/trace/span.h#L32)? This would enforce the spec requirement, I think this is a better and less ambiguous way to add links.

`trace::Link` as it is now is not useful. An API user couldn't even construct a `trace::Link`, as it's a pure interface.

Reply:
@pyohannes  - Sending links via `StartSpanOptions` won't be feasible, as we need to initialize collection of links. Using vector/array will break the ABI. Better option would be to have as separate argument in `StartSpan` : 

```
 nostd::shared_ptr<Span> StartSpan(nostd::string_view /*name*/,
                                    const KeyValueIterable & /*attributes*/,
                                    const StartSpanOptions & /*options*/,
                                    const nostd::span<nostd::shared_ptr<Link>>& /*links*/) noexcept override
```

I don't have strong opinion in having dedicated `Link` type, but since link contains attribute pairs , and so the overloading for `KeyValueIterable` can be done within `Link` type, instead of adding  more overloaded `StartSpan` methods. Let me know what you think ?

Reply:
I like the suggestion of having this as additional argument to the `StartSpan` function. I'm not yet sure about the best way to realize this, but let's work towards that!

Reply:
FYI this is what OpenTelemetry Python has https://github.com/open-telemetry/opentelemetry-python/blob/982b667ab7d32de45ed3b9cd6981f37a9e04975b/opentelemetry-sdk/src/opentelemetry/sdk/trace/__init__.py#L701

Reply:
@pyohannes  - I have done the changes as separate PR #351 , as this was mostly a rework. Please review that PR. I am closing current PR.